keps/sig-storage/1698-generic-inline-volumes/README.md

```diff
@@ -0,0 +1,636 @@
+
+The volume scheduling library and kubelet must check that a PVC is
+owned by the pod before using it. Otherwise they must ignored it. This
+ensures that a volume is not used accidentally for a pod in case of a
+name conflict.
+
+### Pod events
+
+Errors that occur while creating the PVC will be reported as events
+for the pod and thus be visible to the user. This includes quota
+violations and name collisions.
+
+### Feature gate
+
+The `GenericInlineVolumes` feature gate controls whether:
+- the new controller is active in the `kube-controller-manager`,
+- new pods can be created with a `InlineVolumeSource`,
+- anything that specifically acts upon an `InlineVolumeSource` (scheduler,
+  kubelet, etc.) instead of merely copying it (statefulset controller)
+  accepts the new volume source.
+
+Existing pods with such a volumes will not be scheduled respectively
+started when the feature gate is off.
+
+### Modifying volumes
+
+Once the PVC for an inline volume has been created, it can be updated
+directly like other PVCs. The controller will not interfere with that
+because it never updates PVCs. This can be used to control features
+like volume resizing.
+
+### Late binding
+
+Ideally, provisioning should use late binding (aka
+`WaitForFirstConsumer`). The initial implementation assumes that the
+user is taking care of that by selecting that provisioning mode in the
+storage class for the PVC.
+
+Later, `kube-scheduler` and `external-provisioner` can be changed to
+automatically enable late binding for PVCs which are owned by a pod.
+
+### Test Plan
+
+- Unit tests will be added for the API change.
+- Unit tests will cover the functionality of the controller, similar
+  to
+  https://github.com/kubernetes/kubernetes/blob/v1.18.2/pkg/controller/volume/persistentvolume/pv_controller_test.go.
+- Unit tests need to cover the positive case (feature enabled) as
+  well as negative case (feature disabled or feature used incorrectly).
+- A new [storage test
+  suite](https://github.com/kubernetes/kubernetes/blob/2b2cf8df303affd916bbeda8c2184b023f6ee53c/test/e2e/storage/testsuites/base.go#L84-L94)
+  will be added which tests inline volume creation, usage and deletion
+  in combination with all drivers that support dynamic volume
+  provisioning.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Errors emitted as pod events
+- Decide whether `CSIVolumeSource` (in beta at the moment) should be
+  merged with `EphemeralVolumeSource`
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- 3 examples of real world usage
+- Downgrade tests and scalability tests
+- Allowing time for feedback
+
+### Upgrade / Downgrade Strategy
+
+When downgrading to a cluster which does not support generic inline
+volumes (either by disabling the feature flag or an actual version
+downgrade), pods using generic inline volumes will no longer be
```

Comment:
s/inline/ephemeral

Reply:
This one too.

Reply:
Updated. I consider the full name of the feature "generic ephemeral inline volumes". Both are relevant aspects of the volumes, so only mentioning one of them is arbitrary. "Generic" has to be added to distinguish from the existing "CSI ephemeral inline volumes".